crates/runtime/src/databackend/postgres.rs

```diff
@@ -0,0 +1,227 @@
+}
+
+impl DataPublisher for PostgresBackend {
+    fn add_data(&self, _dataset: Arc<Dataset>, data_update: DataUpdate) -> AddDataResult {
+        let pool = Arc::clone(&self.pool);
+        let name = self.name.clone();
+        Box::pin(async move {
+            let mut conn = pool.connect().await.context(DbConnectionPoolSnafu)?;
+            let Some(conn) = conn.as_any_mut().downcast_mut::<PostgresConnection>() else {
+                return Err(
+                    Box::new(Error::UnableToDowncastDbConnection {}) as Box<dyn std::error::Error>
+                );
+            };
+
+            let mut postgres_update = PostgresUpdate {
+                name,
+                data: data_update.data,
+                update_type: data_update.update_type,
+                postgres_conn: conn,
+                create_mutex: &self.create_mutex,
+            };
+
+            postgres_update.update().await?;
+
+            self.initialize_datafusion().await?;
+            Ok(())
+        })
+    }
+
+    fn name(&self) -> &str {
+        "Postgres"
+    }
+}
+
+impl PostgresBackend {
+    #[allow(clippy::needless_pass_by_value)]
+    pub async fn new(
+        ctx: Arc<SessionContext>,
+        name: &str,
+        mode: Mode,
+        params: Arc<Option<HashMap<String, String>>>,
+        primary_keys: Option<Vec<String>>,
+    ) -> Result<Self> {
+        let pool = PostgresConnectionPool::new(name, mode, params)
+            .await
+            .context(DbConnectionPoolSnafu)?;
+        Ok(PostgresBackend {
+            ctx,
+            name: name.to_string(),
+            pool: Arc::new(pool),
+            create_mutex: Mutex::new(()),
+            _primary_keys: primary_keys,
+        })
+    }
+
+    async fn initialize_datafusion(&self) -> Result<()> {
+        let table_exists = self
+            .ctx
+            .table_exist(TableReference::bare(self.name.clone()))
+            .context(DataFusionSnafu)?;
+        if table_exists {
+            return Ok(());
+        }
+
+        let table = match SqlTable::new(&self.pool, TableReference::bare(self.name.clone()))
+            .await
+            .context(PostgresDataFusionSnafu)
+        {
+            Ok(table) => table,
+            Err(e) => return Err(e),
+        };
+
+        self.ctx
+            .register_table(&self.name, Arc::new(table))
+            .context(DataFusionSnafu)?;
+
+        Ok(())
+    }
+}
+
+struct PostgresUpdate<'a> {
+    name: String,
+    data: Vec<RecordBatch>,
+    update_type: UpdateType,
+    postgres_conn: &'a mut dbconnection::postgresconn::PostgresConnection,
+    create_mutex: &'a Mutex<()>,
+}
+
+impl<'a> PostgresUpdate<'a> {
+    async fn update(&mut self) -> Result<()> {
+        if !self.table_exists().await {
+            self.create_table()?;
+        } else if self.update_type == UpdateType::Overwrite {
+            self.postgres_conn
+                .execute(format!(r#"TRUNCATE TABLE "{}""#, self.name).as_str(), &[])
```

Comment:
We should run this update in a transaction - and truncate breaks out of transactions on Postgres, so it would need to be a `DELETE FROM "{}"`